File: solutions/c/complex-numbers/1/complex_numbers.c
#include "complex_numbers.h"
#include <math.h>

// typedef struct {
//    double real;
//    double imag;
// } complex_t;

complex_t c_add(complex_t a, complex_t b)
{
    complex_t res = {a.real + b.real, a.imag + b.imag};
    return res;
}

complex_t c_sub(complex_t a, complex_t b)
{
    complex_t res = {a.real - b.real, a.imag - b.imag};
    return res;
}

complex_t c_mul(complex_t a, complex_t b)
{
    complex_t res = {a.real * b.real - a.imag * b.imag, a.imag * b.real + a.real * b.imag};
    return res;
}

complex_t c_div(complex_t a, complex_t b)
{
    double denom = b.real * b.real + b.imag * b.imag;
    complex_t res = {
        (a.real * b.real + a.imag * b.imag) / denom,
        (a.imag * b.real - a.real * b.imag) / denom
    };
    return res;
}

double c_abs(complex_t x)
{
   return sqrt(x.real * x.real + x.imag * x.imag);
}

complex_t c_conjugate(complex_t x)
{
    complex_t res = {x.real, x.imag * (-1)};
    return res;
}

double c_real(complex_t x)
{
   return x.real;
}

double c_imag(complex_t x)
{
   return x.imag;
}

complex_t c_exp(complex_t x)
{
   complex_t res = {exp(x.real) * cos(x.imag), exp(x.real) * sin(x.imag)};
    return res;
}

File: solutions/go/complex-numbers/1/complex_numbers.go
package complexnumbers

import (
    "math"
)

// Define the Number type here.
type Number struct {
    r float64
    i float64
}

func (n Number) Real() float64 {
	return n.r
}

func (n Number) Imaginary() float64 {
	return n.i
}

func (n1 Number) Add(n2 Number) Number {
	return Number{
        r: n1.r + n2.r,
        i: n1.i + n2.i,
    }
}

func (n1 Number) Subtract(n2 Number) Number {
    return Number{
        r: n1.r - n2.r,
        i: n1.i - n2.i,
    }
}

func (n1 Number) Multiply(n2 Number) Number {
	return Number{
        r: n1.r*n2.r - n1.i*n2.i,
        i: n1.i*n2.r + n1.r*n2.i,
    }
}

func (n Number) Times(factor float64) Number {
	return Number{
        r: n.r * factor,
        i: n.i * factor,
    }
}

func (n1 Number) Divide(n2 Number) Number {
	denom := n2.r*n2.r + n2.i*n2.i
    return Number{
        r: (n1.r*n2.r + n1.i*n2.i) / denom,
        i: (n1.i*n2.r - n1.r*n2.i) / denom,
    }
}

func (n Number) Conjugate() Number {
	return Number{
        r: n.r,
        i: n.i * -1,
    }
}

func (n Number) Abs() float64 {
	return math.Sqrt(n.r * n.r + n.i * n.i)
}

func (n Number) Exp() Number {
	return Number{
        r: math.Exp(n.r) * math.Cos(n.i),
        i: math.Exp(n.r) * math.Sin(n.i),
    }
}
